File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList();
        Set<String> set = new HashSet();
        for(int i=0; i<nums.length-2; i++) {
            int l = i+1;
            int r = nums.length-1;
            
            while(l<r) {
                int sum = nums[i] + nums[l] + nums[r];
                if(sum == 0) {
                    List<Integer> triplet = new ArrayList();
                    triplet.add(nums[i]);
                    triplet.add(nums[l]);
                    triplet.add(nums[r]);
                    Collections.sort(triplet);
                    String tStr = triplet.toString();
                    if(!set.contains(tStr)) {
                        res.add(triplet);
                        set.add(tStr);
                    }
                    l++;
                    while(nums[l] == nums[l-1] && l < r) l++;
                } else if(sum > 0) {
                    r--;
                } else {
                    l++;
                }
            }
        }
        return res;
    }
}

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    const res = []
    nums.sort((a,b) => a-b)
    for(let i=0; i<nums.length; i++) {
        let l = i+1, r=nums.length-1
        if(i > 0 && nums[i] == nums[i-1]) continue
        while(l < r) {
            const sum = nums[i] + nums[l] + nums[r]
            if(sum === 0) {
                res.push([nums[i], nums[l], nums[r]])
                l++
                while(nums[l] === nums[l-1] && l<r) {
                    l++
                }
            } else if(sum > 0) {
                r -= 1
            } else {
                l +=1
            }
        }
    }
    return res
};